agent/grpc-external/limiter/limiter.go

```diff
@@ -0,0 +1,254 @@
+// eventually be drained.
+func (l *SessionLimiter) hasCapacity() bool {
+	max := atomic.LoadUint32(&l.max)
+	if max == Unlimited {
+		return true
+	}
+
+	cur := atomic.LoadUint32(&l.inFlight)
+	return max > cur
+}
+
+// Note: overCapacity is *best effort*. As we do not hold l.mu it's possible that:
+//
+//	- max has changed by the time we compare it to inFlight.
+//	- inFlight > max now, but decreases before we terminate a session.
+func (l *SessionLimiter) overCapacity() bool {
+	max := atomic.LoadUint32(&l.max)
+	if max == Unlimited {
+		return false
+	}
+
+	cur := atomic.LoadUint32(&l.inFlight)
+	return cur > max
+}
+
+func (l *SessionLimiter) terminateSession() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	idx := rand.Intn(len(l.sessionIDs))
+	id := l.sessionIDs[idx]
+	l.sessions[id].terminate()
+	l.deleteSessionLocked(idx)
+}
+
+func (l *SessionLimiter) createSessionLocked() *session {
+	session := &session{
+		l:      l,
+		id:     l.maxSessionID,
+		termCh: make(chan struct{}),
+	}
+
+	l.maxSessionID++
+	l.sessionIDs = append(l.sessionIDs, session.id)
+	l.sessions[session.id] = session
+
+	atomic.AddUint32(&l.inFlight, 1)
+
+	return session
+}
+
+func (l *SessionLimiter) deleteSessionLocked(idx int) {
+	delete(l.sessions, l.sessionIDs[idx])
+
+	// Note: it's important that we preserve the order here (which most allocation
+	// free deletion tricks don't) because we binary search the slice.
+	l.sessionIDs = append(l.sessionIDs[:idx], l.sessionIDs[idx+1:]...)
+
+	atomic.AddUint32(&l.inFlight, ^uint32(0))
+}
+
+// Session allows its holder to perform an operation (e.g. serve a gRPC stream)
+// concurrenly with other session-holders. Sessions may be terminated abruptly
+// by the SessionLimiter, so it is the responsibility of the holder to receive
+// on the Terminated channel and halt the operation when it is closed.
+type Session interface {
+	// End the session.
+	//
+	// This MUST be called when the session-holder is done (e.g. the gRPC stream
+	// is closed).
+	End()
+
+	// Terminated is a channel that is closed when the session is terminated.
+	//
+	// The session-holder MUST receive on it and exit (e.g. close the gRPC stream)
+	// when it is closed.
+	Terminated() <-chan struct{}
+}
+
+type session struct {
+	l *SessionLimiter
+
+	id     uint64
+	termCh chan struct{}
+
+	// done ensures that if both End and terminate are called on a session, only
+	// the first call will have effect - this is important as both will attempt to
+	// clean up the session's state. It also ensures calling End more than once is
+	// a no-op.
+	done uint32
+}
+
+func (s *session) End() {
+	if !atomic.CompareAndSwapUint32(&s.done, 0, 1) {
+		return
+	}
+
+	s.l.mu.Lock()
+	defer s.l.mu.Unlock()
+
+	idx := sort.Search(len(s.l.sessionIDs), func(i int) bool {
+		return s.l.sessionIDs[i] >= s.id
+	})
+
+	if idx == len(s.l.sessionIDs) || s.l.sessionIDs[idx] != s.id {
+		panic(fmt.Sprintf("could not find session id: %d", s.id))
+	}
+
+	s.l.deleteSessionLocked(idx)
```

Comment:
❗ (blocking)

This feels like it should live in a separate method on the SessionLimiter. What do you think? Both are in the same package/file so it would be an incredibly minor difference. It feels a little off for this types methods to be performing the binary search on the internal sessions of the limiter.

Also there seems to be a race condition around looking up the session here? The scenario I am thinking of is:

* The xDS session is ending and it starts to call the `End()` method on the session.
* At the same time the session limiter chooses that session to be terminated. 
   * First it closes the sessions chan
   * Next it gains the lock and deletes the session from tracking
* Now the `End()` method continues execution. The search fails to find the session and panics.

I think the fix would be to just not panic. If the limiter isn't tracking the session its because its already been delete and thus everything is as it should be.

Reply:
Yeah I agree, it does feel a bit funky for `session` to be reaching into `SessionLimiter`'s state - will move it to a method there.

Good catch with the race condition. It took me a little while to grok! I think if we were to remove the `panic` we could also remove the whole `CAS(s.done)` thing too.